rust/xaynet-analytics/src/controller.rs

```diff
@@ -0,0 +1,158 @@
+    const SCREEN_ROUTE_COLLECTION_NAME: &'ctrl str = "screen_routes";
+    const SEND_DATA_FREQUENCY_HOURS: i64 = 24;
+
+    pub fn init(
+        path: String,
+        is_charging: bool,
+        is_connected_to_wifi: bool,
+    ) -> Result<AnalyticsController<'ctrl>, Error> {
+        let analytics_event_repo = AnalyticsEventRepo::new(Self::ANALYTICS_EVENT_COLLECTION_NAME);
+        let controller_data_repo = ControllerDataRepo::new(Self::CONTROLLER_DATA_COLLECTION_NAME);
+        let screen_route_repo = ScreenRouteRepo::new(Self::SCREEN_ROUTE_COLLECTION_NAME);
+        let schemas = vec![
+            AnalyticsEvent::get_schema(Self::ANALYTICS_EVENT_COLLECTION_NAME)?,
+            ControllerData::get_schema(Self::CONTROLLER_DATA_COLLECTION_NAME)?,
+            ScreenRoute::get_schema(Self::SCREEN_ROUTE_COLLECTION_NAME)?,
+        ];
+        let db = IsarDb::new(&path, schemas)?;
+        let last_time_data_sent = Self::get_last_time_data_sent(&controller_data_repo, &db)?;
+        let combiner = DataCombiner {};
+        let sender = Sender {};
+        Ok(AnalyticsController {
+            db,
+            is_charging,
+            is_connected_to_wifi,
+            last_time_data_sent,
+            analytics_event_repo,
+            controller_data_repo,
+            screen_route_repo,
+            combiner,
+            sender,
+            send_data_frequency: Duration::hours(Self::SEND_DATA_FREQUENCY_HOURS),
+        })
+    }
+
+    pub fn dispose(self) -> Result<(), Error> {
+        self.db.dispose()
+    }
+
+    pub fn save_analytics_event(&self, event_bytes: &[u8]) -> Result<(), Error> {
+        let event = AnalyticsEvent::read(event_bytes);
+        self.check_for_new_screen_route(&event)?;
+        self.analytics_event_repo.add(&event, &self.db)
+    }
+
+    pub fn change_connectivity_status(&mut self) {
+        self.is_connected_to_wifi = !self.is_connected_to_wifi
+    }
+
+    pub fn change_state_of_charge(&mut self) {
+        self.is_charging = !self.is_charging
+    }
+
+    pub fn maybe_send_data(&mut self) -> Result<(), Error> {
+        if self.should_send_data() {
+            self.send_data()
+        } else {
+            Ok(())
+        }
+    }
+
+    fn check_for_new_screen_route(&self, event: &AnalyticsEvent) -> Result<(), Error> {
+        match event.screen_route {
+            Some(screen_route) => self.add_screen_route_if_new(screen_route),
+            None => Ok(()),
+        }
+    }
+
+    fn add_screen_route_if_new(&self, screen_route: &ScreenRoute) -> Result<(), Error> {
+        if !self
+            .screen_route_repo
+            .get_all(&self.db)?
+            .contains(screen_route)
+        {
+            self.screen_route_repo.add(screen_route, &self.db)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn get_last_time_data_sent(
+        repo: &ControllerDataRepo,
+        db: &IsarDb,
+    ) -> Result<Option<DateTime<Utc>>, Error> {
+        match repo.get_all(db)?.last() {
```

Comment:
Yes, this is a good point and totally makes sense. i wanted to put it together with vectors because it was easier for me to wrap my head around. the implementation of the iterators totally makes sense, i just would like that to come after all the current pieces being worked on are sawn together, which would be one this PR, the read/get stuff and the unit tests.

As a separate point regarding this specific issue, i haven't decided the design of this data model very well yet. Maybe we want to save a new timestamp (plus other related data we'll find out in the future) every time we send the data. Alternatively, it would be easier to just keep on updating this value and keep it up to date with the latest timestamp. We lose the past data, but it's less wasteful when retrieve it. But anyway, the iterator implementation will definitely help.